drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,338 @@
+
+    uint8_t cfg_a = 0;
+    uint8_t cfg_b = 0;
+    uint8_t mode = 0;
+
+    /* set configuration register A and B */
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MA, params->meas_avg);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_MS, params->meas_mode);
+    _set_reg_bit(&cfg_a, HMC5883L_REG_CFG_A_DO, params->dor);
+    _set_reg_bit(&cfg_b, HMC5883L_REG_CFG_B_GN, params->gain);
+
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_A, &cfg_a, 1), res);
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_B, &cfg_b, 1), res);
+
+    /* set operation mode */
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_HS, 0);
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_MD, params->op_mode);
+
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_MODE, &mode, 1), res);
+
+    /* wait 6 ms accoring to data sheet */
+    xtimer_usleep(6 * US_PER_MS);
+
+    return res;
+}
+
+int hmc5883l_data_ready(const hmc5883l_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    int res = HMC5883L_OK;
+
+    uint8_t reg;
+
+    EXEC_RET(_reg_read(dev, HMC5883L_REG_STATUS, &reg, 1), res);
+    return _get_reg_bit(reg, HMC5883L_REG_STATUS_RDY) ? HMC5883L_OK
+                                                      : HMC5883L_ERROR_NO_DATA;
+}
+
+/*
+ * scale factors for conversion of raw sensor data to degree for possible
+ * sensitivities according to mechanical characteristics in datasheet
+ */
+static const uint32_t HMC5883L_RES[] = {
+     730,      /* uG/LSb for HMC5883L_GAIN_1370 with range +-0.88 Gs */
+     920,      /* uG/LSb for HMC5883L_GAIN_1090 with range  +-1.3 Gs */
+    1220,      /* uG/LSb for HMC5883L_GAIN_820  with range  +-1.9 Gs */
+    1520,      /* uG/LSb for HMC5883L_GAIN_660  with range  +-2.5 Gs */
+    2270,      /* uG/LSb for HMC5883L_GAIN_440  with range  +-4.0 Gs */
+    2560,      /* uG/LSb for HMC5883L_GAIN_390  with range  +-4.7 Gs */
+    3030,      /* uG/LSb for HMC5883L_GAIN_330  with range  +-5.6 Gs */
+    4350,      /* uG/LSb for HMC5883L_GAIN_230  with range  +-8.1 Gs */
+};
+
+int hmc5883l_read(const hmc5883l_t *dev, hmc5883l_data_t *data)
+
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(data != NULL);
+    DEBUG_DEV("data=%p", dev, data);
+
+    int res = HMC5883L_OK;
+
+    hmc5883l_raw_data_t raw;
+
+    EXEC_RET(hmc5883l_read_raw (dev, &raw), res);
+
+    data->x = ((uint32_t)raw.x * HMC5883L_RES[dev->params.gain]);
+    data->y = ((uint32_t)raw.y * HMC5883L_RES[dev->params.gain]);
+    data->z = ((uint32_t)raw.z * HMC5883L_RES[dev->params.gain]);
+
+    return res;
+}
+
+int hmc5883l_read_raw(const hmc5883l_t *dev, hmc5883l_raw_data_t *raw)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(raw != NULL);
+    DEBUG_DEV("raw=%p", dev, raw);
+
+    int res = HMC5883L_OK;
+
+    uint8_t data[6];
+
+    /* read raw data sample */
+    EXEC_RET_CODE(_reg_read(dev, HMC5883L_REG_OUT_X_MSB, data, 6),
+                  res, -HMC5883L_ERROR_RAW_DATA);
+
+    /* data MSB @ lower address */
+    raw->x = (data[0] << 8) | data[1];
```

Comment:
We don't do it any better way in most of the drivers in RIOT, but maybe its a good point to start: using proper byteorder functions here might make this code also runnable on big-endian systems...

Reply:
I'm not very clear, the endianess would be relevant only if `data` would be mapped directly the `int16_t` member. 

`data` contains the bytes in the order as they come from the sensor. This order is always little-endian. `(data[0] << 8) | data[1])` gives always a 16 bit integer which is assigned to the 16 bit integer `raw->x`. The order in which this integer is then stored in memory doesn't matter here. The code should work on little as well as on bigendian. Finally, the code is doing the same as 
`byteorder_bebuftohs` but for integer.

BTW, the byteorder functions in `core/include/bytorder.h` are only defined for unsigned integers which could lead to compiler errors for nitpicking compilers when used with integers.

Reply:
yes, makes sense. I did not look into the semantics of the `raw->x/y/z` fields close enough, so never mind.